src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,555 @@
+
+    /**
+     * @return static
+     */
+    abstract public function beforeLast(string $needle, bool $includeNeedle = false): self;
+
+    /**
+     * @return static[]
+     */
+    abstract public function chunk(int $length = 1): array;
+
+    /**
+     * @return static
+     */
+    public function collapseWhitespace(): self
+    {
+        $str = clone $this;
+        $str->string = trim(preg_replace('/(?:\s{2,}+|[^\S ])/', ' ', $str->string));
+
+        return $str;
+    }
+
+    abstract public function equalsTo(string $string): bool;
+
+    abstract public function endsWith(string $suffix): bool;
+
+    /**
+     * @return static
+     */
+    public function ensureLeft(string $prefix): self
+    {
+        return $this->startsWith($prefix) ? clone $this : $this->prepend($prefix);
+    }
+
+    /**
+     * @return static
+     */
+    public function ensureRight(string $suffix): self
+    {
+        return $this->endsWith($suffix) ? clone $this : $this->append($suffix);
+    }
+
+    public function isEmpty(): bool
+    {
+        return '' === $this->string;
+    }
+
+    public function isUtf8(): bool
+    {
+        return '' === $this->string || preg_match('//u', $this->string);
+    }
+
+    /**
+     * @return static
+     */
+    public function join(array $strings): self
+    {
+        $str = clone $this;
+
+        foreach ($strings as $k => $s) {
+            if ('' !== $s && !preg_match('//u', $s)) {
+                throw new InvalidArgumentException(sprintf('Invalid UTF-8 string at index "%s".', $k));
+            }
+        }
+
+        $str->string = implode($str->string, $strings);
+
+        return $str;
+    }
+
+    abstract public function length(): int;
+
+    public function match(string $pattern, int $flags = 0, int $offset = 0): ?array
+    {
+        if ('g' === $pattern[-1]) {
+            $pattern = substr($pattern, 0, -1);
+            $match = 'preg_match_all';
+        } else {
+            $match = 'preg_match';
+        }
+
+        if ($this->ignoreCase) {
+            $pattern .= 'i';
+        }
+
+        set_error_handler(static function ($t, $m) { throw new InvalidArgumentException($m); });
+
+        try {
+            if (false === $ret = $match($pattern.'u', $this->string, $matches, $flags, $offset)) {
+                $lastError = preg_last_error();
+
+                foreach (get_defined_constants(true)['pcre'] as $k => $v) {
+                    if ($lastError === $v && '_ERROR' === substr($k, -6)) {
+                        throw new RuntimeException('Matching failed with '.$k.'.');
+                    }
+                }
+
+                throw new RuntimeException('Matching failed with unknown error code.');
+            }
+        } finally {
+            restore_error_handler();
+        }
+
+        return $ret ? $matches : null;
+    }
+
+    abstract public function indexOf(string $needle, int $offset = 0): ?int;
```

Comment:
```suggestion
    abstract public function search(string $needle, int $offset = 0): ?int;
```

Reply:
rejected: indexOf is the name almost everyone uses :)